docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+# Advanced Filtering for Triggers Proposal
+
+## Problem
+
+As an `Event Consumer` I want to filter events in triggers with a variety of
+strategies beyond exact match on predefined attributes. Filtering strategies
+that have been requested include:
+
+*   Filtering by source or type prefix or regular expression
+*   Filtering by a custom extension attribute
+*   Filtering by data fields
+
+The existing filter capability of exact match on source and type is simple to
+understand and use, but doesn't support these more advanced filtering scenarios.
+
+## Objective
+
+Design a flexible filtering mechanism that supports all of the advanced use
+cases identified and can also implement simpler cases like exact match.
+
+The filtering mechanism MUST be safe and efficient. It SHOULD be compatible with
+a future
+[upstream propagation solution](https://github.com/knative/eventing/issues/934)
+for delivery efficiency at scale.
+
+Filters should be easy for the user to specify and simple to interpret at a
+glance.
+
+## Requirements
+
+*   Support all of the advanced filtering use cases identified.
+*   Can implement simpler cases like exact match.
+*   Safe and secure when embedded in multi-tenant processes.
+*   Compatible with upstream filter propagation.
+*   Easily specified by the user.
+*   Simple for users to interpret.
+
```

Comment:
What about providing testing tools so users can test changes and not accidentally break their filters?

Reply:
There's a nice cross-platform [conformance test framework](https://github.com/google/cel-spec/blob/master/tests/README.md) for CEL which could be used to test that filters are well-behaved for certain inputs. @kyessenov from the Istio team is currently working on connecting the framework to the C++ evaluator, and the Go stack fully supports the framework and existing tests.

Reply:
@trshafer can you describe in greater detail how you expect this to work? Should it be server-side or client-side?

It might be difficult to define what it means to break a filter. Is a filter broken if the set of events that would previously have passed no longer pass, or new events now pass when they did not before? We don't know whether the user intended that change or not.

If broken means "expression does not compile", we can catch most of those changes at the webhook as described in https://github.com/knative/eventing/pull/1047#discussion_r277449330.

We might even be able to detect issues with dynamic fields by caching information about events that were previously seen by the Trigger and attempting to determine if the new expression still works. However, this is problematic for the same reason as above: the user may intend to evaluate events that do not have the same dynamic fields as the ones the Trigger expects. In fact, this may be quite common since (depending on implementation details) the Trigger filter may evaluate all events coming into the Broker.

There may be an opportunity to leverage the registry for expression validation, if the registry contains canonical event schemas or examples.

If the concern is accidentally dropping events that were in fact interesting to me, there may be other solutions that don't require understanding the semantics of the expression. We could allow the user to specify fail open or fail closed:

```yaml
spec:
  filter:
    cel:
      expression: data.foo != "bar"
      onError: accept
```

Another option is to allow specifying a dead letter queue:

```yaml
spec:
  filter:
    cel:
      expression:
      onError: deadletter
      # optional - there could be a sensible default
      # deadLetterSink:
      #   apiVersion: ...
      #   ...

Reply:
Thanks again for your thoughtful response. In designing a framework I try to think about making the good things easy and the bad things hard. I don't think users are used to writing tests for their k8s yaml because the yaml generally contains configuration and not code. I think relying on users to write a test suite that looks like this could be a difficult ask:

```go
const (
  expression = YAML.parse('mytrigger.yaml').spec.filter.cel.expression
)

func TestInputA(t *testing.T) {
  if (!CEL.evaluate(expression, inputA)) {
    t.Error("Not true");
  }
}
```

Providing a go package that makes testing of the Triggers code easier is probably a good thing. WDYT?

Reply:
> Providing a go package that makes testing of the Triggers code easier is probably a good thing. WDYT?

@trshafer I'm still confused about what you're asking for. The code that reconciles a Trigger and implements the Trigger data plane is written by the Knative _contributors_, not the user (we call this [the _developer_ persona](https://github.com/knative/eventing/blob/master/docs/personas.md)). How would the _developer_ use the go package you propose?

Reply:
Since the yaml will now contain source code, I think it's worth considering how this impacts the developer flow. Developers will be putting code within their k8s yaml which is unlikely to be tested until it gets `kubectl apply`ed to a cluster and is running in production. The go package would make it easier for the developer persona to write test cases against their CEL expressions that are embedded in yaml. If the structure were:

```yaml
spec:
  filter:
    cel:
      expressionFile: some-express-file.cel
``` 
 Then I think it would make less sense to include a trigger test package for yaml. In the file case, a developer could load up the cel file. However since the yaml now contains code, it might make sense to provide a tool or documentation about the developer properly testing their cel expressions for example: https://github.com/google/cel-go#parse-and-check.